### PR TITLE
Meson modules

### PR DESCRIPTION
**Pull request by Saturon**
In complex projects consisting of many different modules, you often have to link modules only with separate libraries from the package (NOT with package at once). In the current implementation, Meson does not allow this.

Therefore added the ability to specify a list of certain libraries for linking. (the modules parameter) This was previously implemented for Boost. The usual behavior of meson before modifications and if you do not specify the modules parameter - add all the found libraries.

`dep = dependency('name', modules: [module1, module2, ...])`

If you only need header files, you must specify an empty list []

`dep_headers = dependency('name', modules: [])`

The exception is the boost for it header files are defined as follows:

`dep_boost_headers = dependency('boost')`

In other cases (other dependencies), this will mean adding all the libraries found. It is advisable not to use due to side effects

IMPORTANT 

If you defined a dependency without specifying the modules parameter in some module (meson.build):

`dep_all_libs = dependency('name')`

Then subsequent attempts to determine the same dependency by limiting the list of libraries to the modules parameter will not work:

`dep_lib1 = dependency ('name', modules: ['lib1']) # Will still connect all libraries from the package, despite the explicit indication to use one lib1 library!`

### CODE REVIEW

**Comment by eli-schwartz** [comment]
It's not obvious to me what this PR is attempting to implement. Maybe it would help to include documentation?

**Comment by Saturon** [comment]
> It's not obvious to me what this PR is attempting to implement. Maybe it would help to include documentation?

How to change reference documentation page? not noticeable at first glance. I will do it

This change adds the ability to include dependencies as in the example below. As you can see only 3 libs will be used.

```
dep_gtest = dependency('gtest', modules: ['gmock', 'gtest', 'gmock_main'], static: true)
deps_external = [dep_gtest]

lib_example_main = static_library(
  'example',
  sources: src_ccfiles,
  cpp_args: cxx_example_flags,
  c_args: c_example_flags,
  include_directories: inc_dirs,
  gnu_symbol_visibility: 'hidden',
  link_with: deps_internal,
  dependencies: deps_external,
  install: true
)
```

**Comment by bonzini** [comment]
So if I understand correctly the idea is to keep all the non-`-l` options, and only the `-l` options that were listed in the modules keyword argument?

**Comment by Saturon** [comment]
> So if I understand correctly the idea is to keep all the non-`-l` options, and only the `-l` options that were listed in the modules keyword argument?

Idea is to make analog to boost modules 
Link: 
https://mesonbuild.com/Dependencies.html#boost

**Comment by eli-schwartz** [comment]
That's a bad comparison then. Because gtest is actually three dependencies:
- gtest_main
- gtest
- gmock

And each can be searched for on their own, and some of them are link dependencies of another -- you can't use gtest_main without gtest or gmock, the dependencies don't decompose.

Modules are intended for limited situations where there's ecosystem handling on top of composable dependencies. Boost doesn't have pkg-config files, but we need to find composable boost libraries, some of which depend on each other, and modules is how we do that. Given `dependency('boost', modules: ['foo', 'bar'])`, we want the dependencies "boost-foo" and "boost-bar", and if boost provided pkg-config files those would be boost-foo.pc and boost-bar.pc.

In fact, that's how it works for `dependency('qt5', modules: ['Core', 'Widgets'])`, which has some qt specific handling and then looks up Qt5Core.pc and Qt5Widgets.pc and uses them as-is.

This doesn't genericize to pkg-config dependencies, which are always a single linkable unit. 

In other words, modules are always a special case that is a holder for one or more linkable dependencies. And pkg-config is always a single linkable dependency.

**Comment by Saturon** [comment]
Is there a way to get debug msg or full logs from powershell? 

I want to check ninja file. For some reason on vc2019x64ninja and vc2019x64vs i have problems. (meanwhile on vc2019arm64ninjacross everything is fine)

This part accured in powershell logs, but it is not enough
```
Running unittests.

Traceback (most recent call last):
File "run_unittests.py", line 47, in <module>
from unittests.linuxliketests import LinuxlikeTests
File "<fstring>", line 1
     (case=)
          ^
SyntaxError: invalid syntax
```

**Comment by eli-schwartz** [comment]
> This part accured in powershell logs, but it is not enough

It is enough. That's attempting to use a python feature introduced in 3.8: https://bugs.python.org/issue36817

But Meson supports 3.7 and some CI machines run on 3.7, so that error message can show up.

**Comment by Saturon** [comment]
> > This part accured in powershell logs, but it is not enough
> 
> It is enough. That's attempting to use a python feature introduced in 3.8: https://bugs.python.org/issue36817
> 
> But Meson supports 3.7 and some CI machines run on 3.7, so that error message can show up.

Thanks, I struggled with this one

**Comment by Saturon** [comment]
can i get permission for the rest of the tests?

**Comment by Saturon** [comment]
@eli-schwartz can you start 2nd part of tests?
+ if i need to explain smth feel free to ask :)

**Comment by eli-schwartz** [comment]
As I said earlier, up above, it's still not clear to me what the use case is -- for your gtest example, the upstream way to do this is `dependency('gmock_main')` since gmock_main already depends on gmock and gtest, and *needs* it. If you just want gmock and gtest, not gmock_main, the correct way to do this is `dependency('gmock')`.

gtest doesn't even link to gmock anyway so not sure how your example works.

**Comment by Saturon** [comment]
Okey

So why are we adding this feature? 
Basically, on my job we have like 30 components +-. Every component has from 1 to 7-10 library's inside. Some libraries conflict with each other. 
So when i link apps, i need to specify what to link. You can't just include whole component, it will conflict . You need to specify exactly what are you linking with what. 

Look at this one. Our Crgtools has 20+ libraries. BUT we can use only 5 becose we made this modules options.
```
...
deps_external += dependency( 'crgtools', modules: ['crg_depcheck_system_core', 'crg_os', 'crg_alg_gost_only', 'gost_crg_calculator', 'gost_crg_calculator_ext'], static: true )

cxxflags = []
deps_internal = []


math_tests_v4 = executable(
  'math_tests_v4',
  sources: src_files,
  cpp_args: cxxflags,
  include_directories: inc_dirs,
  link_args: common_ld_flags,
  link_with: deps_internal,
  dependencies: deps_external,
  install: true
)

...
```

 It works just like boost, you link a lot of components and you can:
1)Put what libraries to link ```dep = dependency('name', modules: [module1, module2, ...])```
2)Put only headers to link ```dep_headers = dependency('name', modules: [])```
3)Put everything ```dep_all_libs = dependency('name')``` (This allows us not breaking builds that work. Works just like before)

**Comment by xclaesse** [comment]
I'm not convinced by this feature, it feels like your project should instead generate separate pkg-config files for each libraries. This PR feels like adding hacks for your personal case. I fail to see how this can be generally useful except for shooting into your feet. In most cases excluding libraries from a pkg-config dependency will result in linker issues, because all libs are supposed to be needed.

**Comment by xclaesse** [comment]
> 2)Put only headers to link `dep_headers = dependency('name', modules: [])`

That you can already do with `dependency('name').partial_dependency(includes: true, compile_args: true)`

**Comment by eli-schwartz** [comment]
> Okey
> 
> So why are we adding this feature?

I didn't originally reject and close this PR because I was asking for more details on how it would be used. I wasn't convinced that it's a feature I would feel comfortable in merging, but I wasn't sure if that was because I misunderstood the intention.

> I'm not convinced by this feature, it feels like your project should instead generate separate pkg-config files for each libraries. 

Yes, I agree. That's what Qt does, that's what gtest/gmock do as well.

> Every component has from 1 to 7-10 library's inside. 

The pkg-config file for a single library can list a `Requires: otherlib` and depend on the pkg-config files for each library it also needs. This means you only need to specify a single component and it recursively adds whichever libraries it needs.

> 3)Put everything dep_all_libs = dependency('name') (This allows us not breaking builds that work. Works just like before)

For this you can also have a `name-all.pc` which can be looked up with `dependency('name-all')`, and which simply depends on the pkg-config files for *every* library.

**Comment by Saturon** [comment]
Look at this example. crgtools.pc is generated by pkg-config. It has a lof of static and shared libraries, some libs are for a test purpose only. See crgtools.pc:

prefix=/root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3
libdir=${prefix}/lib
includedir=${prefix}/include

Name: crgtools
Description: Crg Library
Version: 5.11.10
Libs: -L"${libdir}" -lcalc_crg  -lcrg_alg_ext_only  -lcrg_alg_fips_only  -lcrg_alg_full  -lcrg_alg_gost_only  -lcrg_alg_itcs_only  -lcrg_cstest_leaf_module  -lcrg_cstest_leafsl_module  -lcrg_cstest_main_module  -lcrg_cstest_sub_module-X.X  -lcrg_cstest_subsl_module  -lcrg_depcheck_system_core  -lcrg_os  -lcrg_rfci032  -lcrg_util  -lcrgelf  -lcrgtools  -lfips_crg_calculator  -lgost_crg_calculator  -lgost_crg_calculator_ext  -Wl,-rpath,"${libdir}"
Cflags: -I"${includedir}"

If we declare a dependency such as dependency('crgtools') Meson will make a linker command with all crgtools.pc libs included. See linker_log.txt:

/usr/bin/ld.lld-13" --hash-style=both --eh-frame-hdr -m elf_x86_64 -shared -o libs/cpp/some/libsome.so //root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_alg_ext_only.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_alg_fips_only.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_alg_full.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_alg_gost_only.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_alg_itcs_only.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_cstest_leaf_module.so /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_cstest_leafsl_module.so /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_cstest_main_module.so /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_cstest_sub_module-X.X.so /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_cstest_subsl_module.so /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_depcheck_system_core.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_os.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_rfci032.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrg_util.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrgelf.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libcrgtools.so /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libfips_crg_calculator.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libgost_crg_calculator.a /root/.conan/data/crgtools/5.11.10/User1/release_5_11_x/package/aed8aefced55fc9de349ea5d5ffb2f16df3727d3/lib/libgost_crg_calculator_ext.a

There is an issue for us, because we need specified libs, so we worked on a fix to solve this issue.

[crgtools.pc.txt](https://github.com/mesonbuild/meson/files/10489001/crgtools.pc.txt)
[meson_dep_crgtools_linker_args.txt](https://github.com/mesonbuild/meson/files/10489002/meson_dep_crgtools_linker_args.txt)

**Comment by bonzini** [comment]
It seems to me that this can be handled with a combination of partial_args and find_library.
